R/guides.R

```diff
@@ -59,13 +59,66 @@ collapse_guides <- function(guides) {
   }
   guides
 }
+
+#' @importFrom utils getFromNamespace
+#' @importFrom ggplot2 calc_element
+assemble_guides <- function(guides, position, theme) {
+  # https://github.com/tidyverse/ggplot2/blob/57ba97fa04dadc6fd73db1904e39a09d57a4fcbe/R/guides-.R#L512
+  theme$legend.spacing <- theme$legend.spacing %||% unit(0.5, "lines")
+  theme$legend.spacing.y <- calc_element("legend.spacing.y", theme)
+  theme$legend.spacing.x <- calc_element("legend.spacing.x", theme)
+
+  # for every position, collect all individual guides and arrange them
+  # into a guide box which will be inserted into the main gtable
+  Guides <- getFromNamespace("Guides", "ggplot2")
+  package_box <- .subset2(Guides, "package_box") %||% package_box
+  package_box(guides, position, theme)
+}
+
+#' @importFrom grid valid.just editGrob
+package_box <- function(guides, guide_pos, theme) {
+  theme <- complete_guide_theme(guide_pos, theme)
+  guides <- guides_build(guides, theme)
+
+  # Set the justification of the legend box
+  # First value is xjust, second value is yjust
+  just <- valid.just(calc_element("legend.justification", theme))
+  xjust <- just[1]
+  yjust <- just[2]
+  guides <- editGrob(guides,
+    vp = viewport(x = xjust, y = yjust, just = c(xjust, yjust))
+  )
+  guides <- gtable_add_rows(guides, unit(yjust, "null"))
+  guides <- gtable_add_rows(guides, unit(1 - yjust, "null"), 0)
+  guides <- gtable_add_cols(guides, unit(xjust, "null"), 0)
+  guides <- gtable_add_cols(guides, unit(1 - xjust, "null"))
+  guides
+}
+
+#' @importFrom ggplot2 calc_element
+complete_guide_theme <- function(guide_pos, theme) {
+  if (guide_pos %in% c("top", "bottom")) {
+    theme$legend.box <- calc_element("legend.box", theme) %||% "horizontal"
+    theme$legend.direction <- calc_element("legend.direction", theme) %||%
+      "horizontal"
+    theme$legend.box.just <- calc_element("legend.box.just", theme) %||%
+      c("center", "top")
+  } else {
+    theme$legend.box <- calc_element("legend.box", theme) %||% "vertical"
+    theme$legend.direction <- calc_element("legend.direction", theme) %||%
+      "vertical"
+    theme$legend.box.just <- calc_element("legend.box.just", theme) %||%
+      c("left", "top")
+  }
+  theme
+}
+
```

Comment:
Can you please move these functions back to their original position so I can more easily see what has changed?